File: comm/UnionFind.rs
/**
 * Union-Find tree.
 * Verified by https://atcoder.jp/contests/keyence2019/submissions/4011936
 */
struct UnionFind { disj: Vec<usize>, rank: Vec<usize> }

impl UnionFind {
    fn new(n: usize) -> Self {
        let mut disj = vec![0; n];
        for i in 0 .. n {
            disj[i] = i;
        }
        UnionFind { disj: disj, rank: vec![0; n] }
    }
    fn root(self: &mut Self, x: usize) -> usize {
        if x != self.disj[x] {
            let par = self.disj[x];
            let r = self.root(par);
            self.disj[x] = r;
        }
        return self.disj[x];
    }
    fn unite(self: &mut Self, x: usize, y: usize) {
        let mut xr = self.root(x);
        let mut yr = self.root(y);
        if self.rank[xr] > self.rank[yr] {
            std::mem::swap(&mut xr, &mut yr);
        }
        self.disj[xr] = yr;
        self.rank[yr] = std::cmp::max(self.rank[yr], self.rank[xr] + 1);
    }
    fn is_same_set(self: &mut Self, x: usize, y: usize) -> bool {
        return self.root(x) == self.root(y);
    }
}

File: comm/graph/UnionFind.scala
/**
 * Union-Find tree.
 * toArray returns an array consisting of the roots of all vertices.
 * Verified by: AtCoder ARC032-B(http://arc032.contest.atcoder.jp/submissions/603918)
 */
class UnionFind(n: Int) {
  private[this] val disj = Array.tabulate(n)(x => x)
  def root(x: Int): Int =
    if (disj(x) != x) {
      disj(x) = root(disj(x))
      disj(x)
    } else x
  def unite(x: Int, y: Int) {
    disj(root(y)) = root(x)
  }
  def toArray: Array[Int] = {
    for (i <- 0 until n) root(i)
    disj.toArray
  }
}
